File: dev/requirements/logic/cases/audience-descriptors.1.5.case.swift
import Foundation
import SetlistCore

/// 1.5 — A show carries venue, start time, planned set length and structured
/// audience descriptors, each independently unknown.
let audienceDescriptors_1_5 = RequirementCase(id: "1.5") {
    let show = Show(
        venue: "Corp gig, Leith",
        startedAt: REFERENCE_NOW,
        plannedLength: 1200,
        audience: Audience(size: 80, seating: .seated, service: .dinner, crowd: .corporate)
    )
    try expectEqual(show.venue, "Corp gig, Leith", "venue")
    try expectEqual(show.startedAt, REFERENCE_NOW, "start time")
    try expectEqual(show.plannedLength, 1200, "planned length")
    try expectEqual(show.audience.size, 80, "audience size")
    try expectEqual(show.audience.seating, .seated, "seating")
    try expectEqual(show.audience.service, .dinner, "service")
    try expectEqual(show.audience.crowd, .corporate, "crowd")

    // Each descriptor is independently unknown…
    let partial = Audience(seating: .standing)
    try expectNil(partial.size, "unknown size")
    try expectEqual(partial.seating, .standing, "known seating beside an unknown size")

    // …and unknown is not a zero: both states survive a round-trip distinctly.
    let empty = try JSONDecoder().decode(Audience.self, from: Data("{}".utf8))
    try expectNil(empty.size, "absent size decodes as unknown")
    let zero = try JSONDecoder().decode(Audience.self, from: Data(#"{"size":0}"#.utf8))
    try expectEqual(zero.size, 0, "a real zero decodes as zero")
}

File: dev/requirements/logic/cases/audience-descriptors.1.5.case.dart
import 'dart:convert';

import 'package:setlist_core/setlist_core.dart';

import '../../shared/expectations.dart';
import '../../shared/requirement_case.dart';

/// 1.5 — A show carries structured audience descriptors, each independently
/// unknown.
final audienceDescriptors_1_5 = RequirementCase('1.5', (tester) {
  const partial = Audience(seating: Seating.seated);

  expectAbsent(partial.size, 'an unset size');
  expectAbsent(partial.service, 'an unset service');
  expectAbsent(partial.crowd, 'an unset crowd');
  expectThat(partial.seating, Seating.seated, 'while the one that is set is kept');

  // "unknown size, seated" and "0 people, seated" are different facts, and the
  // analysis product would draw different conclusions from them.
  expectFalse(
    partial == const Audience(size: 0, seating: Seating.seated),
    'an unknown size is not an empty room',
  );

  final back = Audience.fromJson(
    json.decode(json.encode(const Audience(size: 80, crowd: Crowd.publicRoom).toJson()))
        as Map<String, dynamic>,
  );
  expectThat(back.size, 80, 'a known descriptor survives the round trip');
  expectAbsent(back.seating, 'and an absent one does not become a default');

  expectAbsent(
    Audience.fromJson(const {'seating': 'hammocks'}).seating,
    'a value this build does not recognise is unknown, not the first option',
  );
});
